ocaml/typing/jkind_types.mli

```diff
@@ -0,0 +1,224 @@
+  | Type_argument of
+      { parent_path : Path.t;
+        position : int;
+        arity : int
+      }
+  | Tuple
+  | Row_variable
+  | Polymorphic_variant
+  | Arrow
+  | Tfield
+  | Tnil
+  | First_class_module
+  | Separability_check
+  | Univar
+  | Polymorphic_variant_field
+  | Default_type_jkind
+  | Existential_type_variable
+  | Array_comprehension_element
+  | Lazy_expression
+  | Class_type_argument
+  | Class_term_argument
+  | Structure_element
+  | Debug_printer_argument
+  | V1_safety_check
+  | Captured_in_object
+  | Recmod_fun_arg
+  | Unknown of string
+
+type immediate_creation_reason =
+  | Empty_record
+  | Enumeration
+  | Primitive of Ident.t
+  | Immediate_polymorphic_variant
+
+type immediate64_creation_reason = Separability_check
+
+type void_creation_reason = |
+
+type any_creation_reason =
+  | Missing_cmi of Path.t
+  | Initial_typedecl_env
+  | Dummy_jkind
+  | Type_expression_call
+  | Inside_of_Tarrow
+  | Wildcard
+  | Unification_var
+  | Array_type_argument
+
+type float64_creation_reason = Primitive of Ident.t
+
+type word_creation_reason = Primitive of Ident.t
+
+type bits32_creation_reason = Primitive of Ident.t
+
+type bits64_creation_reason = Primitive of Ident.t
+
+type annotation_context =
+  | Type_declaration of Path.t
+  | Type_parameter of Path.t * string option
+  | Newtype_declaration of string
+  | Constructor_type_parameter of Path.t * string
+  | Univar of string
+  | Type_variable of string
+  | Type_wildcard of Location.t
+  | With_error_message of string * annotation_context
+
+type creation_reason =
+  | Annotated of annotation_context * Location.t
+  | Missing_cmi of Path.t
+  | Value_creation of value_creation_reason
+  | Immediate_creation of immediate_creation_reason
+  | Immediate64_creation of immediate64_creation_reason
+  | Void_creation of void_creation_reason
+  | Any_creation of any_creation_reason
+  | Float64_creation of float64_creation_reason
+  | Word_creation of word_creation_reason
+  | Bits32_creation of bits32_creation_reason
+  | Bits64_creation of bits64_creation_reason
+  | Concrete_creation of concrete_jkind_reason
+  | Imported
+  | Imported_type_argument of
+      { parent_path : Path.t;
+        position : int;
+        arity : int
+      }
+  (* [position] is 1-indexed *)
+  | Generalized of Ident.t option * Location.t
+
+type interact_reason =
+  | Gadt_equation of Path.t
+  | Tyvar_refinement_intersection
+  | Subjkind
+
+type const =
+  | Any
+  | Value
+  | Void
+  | Immediate64
+  | Immediate
+  | Float64
+  | Word
+  | Bits32
+  | Bits64
+  | Non_null_value
+
+type 'type_expr history =
+  | Interact of
+      { reason : interact_reason;
+        lhs_jkind : 'type_expr Jkind_desc.t;
+        lhs_history : 'type_expr history;
+        rhs_jkind : 'type_expr Jkind_desc.t;
+        rhs_history : 'type_expr history
+      }
+  | Creation of creation_reason
+
+type 'type_expr t =
+  { jkind : 'type_expr Jkind_desc.t;
+    history : 'type_expr history
```

Comment:
I remember why I didn't do this now - the consequence of this is types like `Types.type_expr`, etc needing to also be polymorphic over `history`. Although I agree the duplication is a pain-point, I think it might be the lesser of whatever other evil we do to work around this.

Reply:
Blargh.

Yes I agree with you.

I couldn't take the duplication. So I found another way to get rid of it; see commit on the way (which I would appreciate review on).